docs/tpm-spi-dev.md

```diff
@@ -0,0 +1,368 @@
+API, except for GPIO chip select, which may not be usable from within SPI slave.
+
+### Summary of the HAL-based approach
+
+The STM32 HAL approach was continued as it was the easiest path forward for
+idea verification. Unfortunataly, we faced several obstacles and decided not to
+continue this path, as this approach is not the target one for multiple
+reasons:
+- the STM32L476 is
+  [short on memory](https://lpntpm.lpnplant.io/issues/#memory-usage) when
+  running the `ms-tpm-20-ref` stack
+- the STM32L476 will not be able to provide LPC interface, which is common
+  interface for TPM modules
+- due to the global supply chain issues, the STM32L476 (as most of the others
+  STM32, and many more MCUs) is infeasible to source
+
+Some more considerations on the hardware replaceement could be found
+[here](https://lpntpm.lpnplant.io/further_project_development/).
+
+One of the path to move forward is to port the `ms-tpm-20-ref` stack to some
+RTOS (such as [Zephyr](https://zephyrproject.org/)) to gain some more
+flexibility in tems of hardware switching in the current global situation.
+
+## Prototyping TPM SPI communication with Zephyr
+
+Further SPI work has been done on Zephyr, the code is available
+[here](https://github.com/lpn-plant/zephyr-spi-app). This code tests
+communication with a dummy TPM implementation as `ms-tpm-ref-20` has not been
+fully ported to Zephyr yet.
+
+Zephyr has it's own SPI driver, which still uses low level HAL API, however its
+behaviour is different from high level HAL API we were using previously. First,
+receiving of multiple bytes at once works fine, which simplifies code - instead
+of filling RX buffer one byte at a time we can do this in 2 transfers
+
+- 4 byte TPM header which contains: mode (read or write),
+  target register address and transfer size
+- data read or write transfer with matching size as defined in TPM header
+
+> Note: Current SPI implementation uses different header format than defined by
+> TPM spec. Mode is not present and, on read,  TPM always sends full register,
+> ignoring header size field.
+
+| Byte | Usage            |
+| ---- | ---------------- |
+| 0    | Size of transfer |
+| 1    | Register address |
+| 2    | Unused (zero)    |
+| 3    | Unused (zero)    |
+
+Currently, there are 3 registers (mininum required to send command and receive
+response):
+
+- TPM_REG_CMD - register used for sending command, there is no FIFO, entire
+  command is executed immediatelly after transfer is complete
+- TPM_REG_STATUS - status register, currently contains only response size
+- TPM_REG_RESP - response, there is no FIFO so response must be read during a
+  single transfer
+
+## Running Zephyr sample
+
+Zephyr sample communicates SPI1 (there problems in getting SPI2 to work under
+Zephyr). Following pins are used:
+
+| Pin | Function |
+| --- | -------- |
+| PA5 | SCK      |
+| PA6 | MISO     |
+| PA7 | MOSI     |
+
+> Note: Zephyr uses GPIO chip select on SPI1 which works only for master, not
```

Comment:
What does it mean? The CS is not used? How do we know when we should listen on the bus? Always?

Reply:
Currently, yes, device listens on the bus at all times.

I added more information to this note.